### ISSUE TITLE
Add custom types to default_collate

### ISSUE THREAD

**Issue by 0phoff** [pytorch/pytorch#52353]
## 🚀 Feature
It would be really handy to be able to add custom types to the default_collate function and that everything still works as expected when returning dicts,lists,namedtuples etc.

## Motivation
If you want to accept different types from your dataset than the ones expected by PyTorch, you can do so by providing your own custom `collate_fn`. This is great, but often times you want to simply add a type and still be able to use the types provided by `default_collate`.

One would think that it shouldn't be difficult to create your own wrapper and call the `default_collate` as last resort, thus allowing the use of the default types. However, this does not work for container types.
The current collate implementation will call itself on the transposed list when it encounters one of the list types and that's when your custom collate function (that calls default_collate) breaks, because in this inner function call, your custom type is not supported anymore.

A small diagram to clearly show the issue:
```
batch = [(Tensor, CustomType), (Tensor, CustomType), ...]
custom_collate(batch)
  -> batch is not CustomType:
    -> default_collate(batch)
      -> batch is collections.abc.Sequence
        -> default_collate([Tensor, Tensor, ...])
          -> Tensor
        -> default_collate([CustomType, CustomType, ...])
          -> RuntimeError
```

In order to fix this error you need to copy and paste the following code in each of your custom_collate functions:
``` python
if isinstance(elem, CustomType):
  # Case specific code
  pass
elif isinstance(elem, collections.abc.Mapping):
  return {key: custom_collate([d[key] for d in batch]) for key in elem}
elif isinstance(elem, tuple) and hasattr(elem, '_fields'):  # namedtuple
  return elem_type(*(custom_collate(samples) for samples in zip(*batch)))
elif isinstance(elem, collections.abc.Sequence):
  # check to make sure that the elements in batch have consistent size
  it = iter(batch)
  elem_size = len(next(it))
  if not all(len(elem) == elem_size for elem in it):
    raise RuntimeError('each element in list of batch should be of equal size')
  transposed = zip(*batch)
  return [custom_collate(samples) for samples in transposed]
else:
  return default_collate(batch)
```

## Pitch
I would love to be able to extend the default_collate, without needing to copy and paste the code from default_collate each time.
This would make it easier to extend functionality and thus allow you to close future issues by telling people to implement it themselves (instead of dissapointing them by saying out-of-scope :stuck_out_tongue_winking_eye:).

You could transform the default_collate to a callable class:
```python
class DefaultCollate:
  def __call__(self, batch):
     pass

# Custom collates
class CustomCollate(DefaultCollate):
  def __call__(self, batch):
    if isinstance(batch[0], CustomType):
      pass
    else:
      return super().__call__(batch)
```

## Alternatives
Transforming CustomCollate to a class, just for this feature might be overkill as we don't need to store any state.
Basically, we just need a way to issue some kind of virtual function lookup, in order to call the correct collate_function when recursively collating. Using a class is the first thing that popped into my mind, but maybe there are other solutions ?

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/33181https://github.com/pytorch/pytorch/issues/1512

**Comment by 0phoff** [comment]
> Related: #33181 #1512

Sorry, I apparently did not search well at all, as it seems that #41239 proposes the same thing as me and #33181 would seem to solve this as well!
My bad, closing this !
